python/tests/test_file_echo_api.py

```diff
@@ -306,45 +306,114 @@ def test_caching(self):
 
 
 class RemoteSocketProcessTests(GenericFileEchoTests, unittest.TestCase):
+    # Connection to cryptol
+    c = None
+    # process running the server
+    p = None
+
+    @classmethod
+    def setUpClass(self):
+        p = subprocess.Popen(
+            ["cabal", "run", "file-echo-api", "--verbose=0", "--", "socket", "--port", "50005"],
+            stdout=subprocess.PIPE,
+            stdin=subprocess.DEVNULL,
+            stderr=subprocess.PIPE,
+            start_new_session=True)
+        time.sleep(3)
+        assert(p is not None)
+        poll_result = p.poll()
+        if poll_result is not None:
+            print(poll_result)
+            print(p.stdout.read())
+            print(p.stderr.read())
+        assert(poll_result is None)
+
+        self.p = p
+        self.c = argo.ServerConnection(argo.RemoteSocketProcess('localhost', 50005, ipv6=True))
+
+
+    @classmethod
+    def tearDownClass(self):
+        os.killpg(os.getpgid(self.p.pid), signal.SIGKILL)
+        super().tearDownClass()
+
 
     # to be implemented by classes extending this one
     def get_connection(self):
-        return argo.ServerConnection(
-                argo.RemoteSocketProcess('localhost', 50005, ipv6=True))
+        return self.c
 
     def get_caching_iterations(self):
-        return 100
+        return 30
 
 class DynamicSocketProcessTests(GenericFileEchoTests, unittest.TestCase):
+    # Connection to server
+    c = None
+
+    @classmethod
+    def setUpClass(self):
+        self.c = argo.ServerConnection(
+                    argo.DynamicSocketProcess("cabal v2-exec file-echo-api --verbose=0 -- socket --port 50006"))
 
     # to be implemented by classes extending this one
     def get_connection(self):
-        return argo.ServerConnection(
-                argo.DynamicSocketProcess("cabal v2-exec file-echo-api --verbose=0 -- socket --port 50005"))
+        return self.c
 
     def get_caching_iterations(self):
-        return 100
+        return 30
 
 class StdIOProcessTests(GenericFileEchoTests, unittest.TestCase):
+    # Connection to server
+    c = None
+
+    @classmethod
+    def setUpClass(self):
+        self.c = argo.ServerConnection(
+                    argo.StdIOProcess("cabal v2-exec file-echo-api --verbose=0 -- stdio"))
 
-    # to be implemented by classes extending this one
     def get_connection(self):
-        return argo.ServerConnection(
-                argo.StdIOProcess("cabal v2-exec file-echo-api --verbose=0 -- stdio"))
+        return self.c
 
     def get_caching_iterations(self):
-        return 100
+        return 30
 
 
 class HttpTests(GenericFileEchoTests, unittest.TestCase):
+    # Connection to server
+    c = None
+    # process running the server
+    p = None
+
+    @classmethod
+    def setUpClass(self):
+        p = subprocess.Popen(
+            ["cabal", "run", "file-echo-api", "--verbose=0", "--", "http", "/", "--port", "8080"],
```

Comment:
It's curious that other parts of this file use `cabal exec file-echo-api`, but this part uses `cabal run file-echo-api` instead. Is there a reason for doing something different here?

Reply:
I've preferred `run` over `exec` after getting burned from the latter not taking into account whether things need rebuilt, etc, while the former does. Do you have a strong opinion and/or recommendation either way? I'd be happy to tweak them all to be whatever is "ideal" =)

Reply:
I guess to directly answer your question, had I remembered I would have changed them all to use `run` to fit my personal preferences and avoid confusion down the line for people who run the tests and may have forgotten to explicitly rebuilt the Haskell executable.

Reply:
I don't have a particularly strong opinion one way or the other, provided that you're happy. I'm mainly pointing this out since different tests in this file are now using `run` (e.g., `RemoteSocketProcessTests`) versus `exec` (e.g., `DynamicSocketProcessTests`) inconsistently.

If you do want to use `cabal run`, make sure you have a way to ignore the output from `cabal` itself so that it doesn't mistakenly get recognized as test output.

Reply:
They now match  - if you think it would be better to not use `run` I'm all ears and happy to change things.

Reply:
If `cabal run` works, then I'm quite happy to use that.